// ngx-masonry/esm2015/lib/ngx-masonry.directive.js
import { __decorate, __param } from "tslib";
import { Directive, Inject, ElementRef, forwardRef, OnDestroy, AfterViewInit, Renderer2, OnInit } from '@angular/core';
import { style, animate, AnimationBuilder } from '@angular/animations';
import { NgxMasonryComponent } from './ngx-masonry.component';
let NgxMasonryDirective = class NgxMasonryDirective {
    constructor(builder, element, parent, renderer) {
        this.builder = builder;
        this.element = element;
        this.parent = parent;
        this.renderer = renderer;
        this.animations = {
            show: [
                style({ opacity: 0 }),
                animate('400ms ease-in', style({ opacity: 1 })),
            ],
            hide: [
                style({ opacity: '*' }),
                animate('400ms ease-in', style({ opacity: 0 })),
            ]
        };
    }
    ngOnInit() {
        if (this.parent.options.animations !== undefined) {
            this.animations = this.parent.options.animations;
        }
    }
    ngAfterViewInit() {
        const images = this.element.nativeElement.getElementsByTagName('img');
        this.renderer.setStyle(this.element.nativeElement, 'opacity', '0');
        this.images = new Set(images);
        if (images.length === 0) {
            this.parent.add(this.element.nativeElement);
        }
        else {
            for (const imageRef of images) {
                this.renderer.listen(imageRef, 'load', _ => {
                    this.imageLoaded(imageRef);
                });
                this.renderer.listen(imageRef, 'error', _ => {
                    this.imageLoaded(imageRef);
                });
            }
        }
    }
    ngOnDestroy() {
        if (this.images.size === 0 && this.element.nativeElement.parentNode) {
            this.playAnimation(false);
            this.parent.remove(this.element.nativeElement);
        }
    }
    imageLoaded(image) {
        this.images.delete(image);
        if (this.images.size === 0) {
            this.renderer.setStyle(this.element.nativeElement, 'opacity', '100');
            this.parent.add(this.element.nativeElement);
            this.playAnimation(true);
        }
    }
    playAnimation(show) {
        const metadata = show ? this.animations.show : this.animations.hide;
        if (metadata) {
            const player = this.builder.build(metadata).create(this.element.nativeElement);
            player.play();
        }
    }
};
NgxMasonryDirective.ctorParameters = () => [
    { type: AnimationBuilder },
    { type: ElementRef },
    { type: NgxMasonryComponent, decorators: [{ type: Inject, args: [forwardRef(() => NgxMasonryComponent),] }] },
    { type: Renderer2 }
];
NgxMasonryDirective = __decorate([
    Directive({
        selector: '[ngxMasonryItem], ngxMasonryItem'
    }),
    __param(2, Inject(forwardRef(() => NgxMasonryComponent)))
], NgxMasonryDirective);
export { NgxMasonryDirective };
